executor/write.go

```diff
@@ -65,50 +70,50 @@ func updateRecord(ctx context.Context, h int64, oldData, newData []types.Datum,
 			}
 			t.RebaseAutoID(val, true)
 		}
-		casted, err := table.CastValue(ctx, newData[i], col.ToInfo())
+		cmp, err := newData[i].CompareDatum(sc, oldData[i])
 		if err != nil {
 			return false, errors.Trace(err)
 		}
-		newData[i] = casted
-		touched[i] = true
-		assignExists = true
-	}
-
-	// If no assign list for this table, no need to update.
-	if !assignExists {
-		return false, nil
+		if cmp != 0 {
+			changed = true
+			touched[i] = true
+			if col.IsPKHandleColumn(t.Meta()) {
+				handleChanged = true
+			}
+		} else {
+			if mysql.HasOnUpdateNowFlag(col.Flag) && touched[i] {
```

Comment:
the `touched` is used to update secondary indices if need, so if a field is specified in `set list`, but the value has not changed really, we should not set `touched[that_field]` to `true`.